app/services/warm_cache.rb

```diff
@@ -0,0 +1,36 @@
+class WarmCache
+  def call
+    r = {}
+    # Get all provider who have an api token that has been used in the past 6.months
+    Provider.joins(:vendor_api_tokens)
+      .where('vendor_api_tokens.last_used_at > ?', 1.month.ago).distinct.each do |provider|
+      Rails.logger.tagged('Warm Cache').info "Provider cached: #{provider.id}"
+      # Find the api version they last used
+      vendor_version = VendorAPIRequest
+        .where(provider_id: provider.id)
+        .select("regexp_matches(request_path, '/api/v(.*)/applications') result")
+        .order(created_at: :desc)
+        .first&.result&.first
+
+      # skip as we couldn't find any api requests
+      next unless vendor_version
+
+      r[provider.name] = []
+
+      # For each of the application choices
+      # generate a cache entry
+      ApplicationChoice.joins(course_option: { course: :provider })
```

Comment:
I wonder if we should get the application choices by using this method https://github.com/DFE-Digital/apply-for-teacher-training/blob/53fcaf5fdcdca6cd2723728f51f1e0f7cab1e4c6/app/controllers/concerns/vendor_api/application_data_concerns.rb#L21-L27

In case the method changes the cache warm-up query will change too. Will need to include the concern here and set some private methods. Just a thought.

Reply:
Great suggestion